pkg/envinfo/storage.go

```diff
@@ -56,39 +56,63 @@ func (ei *EnvInfo) GetStorage(name string) (*localConfigProvider.LocalStorage, e
 
 // CreateStorage sets the storage related information in the local configuration
 func (ei *EnvInfo) CreateStorage(storage localConfigProvider.LocalStorage) error {
+	//initialize volume mount and volume container
+	vm := []devfilev1.VolumeMount{
+		{
+			Name: storage.Name,
+			Path: storage.Path,
+		},
+	}
+	vc := []devfilev1.Component{{
+		Name: storage.Name,
+		ComponentUnion: devfilev1.ComponentUnion{
+			Volume: &devfilev1.VolumeComponent{
+				Volume: devfilev1.Volume{
+					Size: storage.Size,
+				},
+			},
+		},
+	}}
+	volumeExists := false
 	// Get all the containers in the devfile
 	containers, err := ei.GetContainers()
 	if err != nil {
 		return err
 	}
 
-	// Add volumeMount to all containers in the devfile
+	// Add volumeMount to all containers if no container is specified else to specified container(s) in the devfile
 	for _, c := range containers {
-		if err := ei.devfileObj.Data.AddVolumeMounts(c.Name, []devfilev1.VolumeMount{
-			{
-				Name: storage.Name,
-				Path: storage.Path,
-			},
-		}); err != nil {
-			return err
+		if storage.Container == "" || (storage.Container != "" && c.Name == storage.Container) {
+			if err := ei.devfileObj.Data.AddVolumeMounts(c.Name, vm); err != nil {
+				return err
+			}
 		}
 	}
 
-	// Add volume component to devfile. Think along the lines of a k8s pod spec's volumeMount and volume.
-	err = ei.devfileObj.Data.AddComponents([]devfilev1.Component{{
-		Name: storage.Name,
-		ComponentUnion: devfilev1.ComponentUnion{
-			Volume: &devfilev1.VolumeComponent{
-				Volume: devfilev1.Volume{
-					Size: storage.Size,
-				},
-			},
-		},
-	}})
+	// Get all the components to check if volume component exists
+	components, err := ei.devfileObj.Data.GetComponents(common.DevfileOptions{})
 	if err != nil {
 		return err
 	}
 
+	// check if volume component already exists
+	for _, component := range components {
+		if component.Volume != nil && component.Name == storage.Name {
+			volumeExists = true
+		}
+	}
+
+	// Add volume component to devfile. Think along the lines of a k8s pod spec's volumeMount and volume.
+	// Add if volume does not exist, otherwise update
+	if !volumeExists {
+		err = ei.devfileObj.Data.AddComponents(vc)
+		if err != nil {
+			return err
+		}
+	} else {
+		ei.devfileObj.Data.UpdateComponent(vc[0])
+	}
```

Comment:
We don't need to update the container as the having a volume based container with the same name will error out in validation. 
https://github.com/openshift/odo/blob/cdd92ef3a5051f340df6427aadc46c17d6a99309/pkg/envinfo/storage.go#L35
We should throw an error here instead.

Reply:
Updated else to return error, will push shortly

Reply:
Updated else to return error